i2c_manager/i2c_manager.c

```diff
@@ -0,0 +1,368 @@
+	#elif defined(I2C_ONE)
+		#define I2C_PORT_CHECK(port, fail) \
+			if (port != I2C_NUM_1) ERROR_PORT(port, fail);
+	#else
+		#define I2C_PORT_CHECK(port, fail) \
+			ERROR_PORT(port, fail);
+	#endif
+#endif
+
+static void i2c_send_address(i2c_cmd_handle_t cmd, uint16_t addr, i2c_rw_t rw) {
+	if (addr & I2C_ADDR_10) {
+		i2c_master_write_byte(cmd, 0xF0 | ((addr & 0x3FF) >> 7) | rw, ACK_CHECK_EN);
+		i2c_master_write_byte(cmd, addr & 0xFF, ACK_CHECK_EN);
+	} else {
+		i2c_master_write_byte(cmd, (addr << 1) | rw, ACK_CHECK_EN);
+	}
+}
+
+static void i2c_send_register(i2c_cmd_handle_t cmd, uint32_t reg) {
+	if (reg & I2C_REG_16) {
+	    i2c_master_write_byte(cmd, (reg & 0xFF00) >> 8, ACK_CHECK_EN);
+	}
+    i2c_master_write_byte(cmd, reg & 0xFF, ACK_CHECK_EN);
+}
+
+esp_err_t I2C_FN(_init)(i2c_port_t port) {
+
+	I2C_PORT_CHECK(port, ESP_FAIL);
+
+	esp_err_t ret = ESP_OK;
+
+	if (I2C_FN(_mutex)[port] == 0) {
+
+		ESP_LOGI(TAG, "Starting I2C master at port %d.", (int)port);
+
+		I2C_FN(_mutex)[port] = xSemaphoreCreateMutex();
+
+		i2c_config_t conf = {0};
+		
+		#ifdef HAS_CLK_FLAGS
+			conf.clk_flags = 0;
+		#endif
+
+		#if defined (I2C_ZERO)
+			if (port == I2C_NUM_0) {
+				conf.sda_io_num = CONFIG_I2C_MANAGER_0_SDA;
+				conf.scl_io_num = CONFIG_I2C_MANAGER_0_SCL;
+				conf.sda_pullup_en = I2C_MANAGER_0_PULLUPS ? GPIO_PULLUP_ENABLE : GPIO_PULLUP_DISABLE;
+				conf.scl_pullup_en = conf.sda_pullup_en;
+				conf.master.clk_speed = CONFIG_I2C_MANAGER_0_FREQ_HZ;
+			}
+		#endif
+
+		#if defined (I2C_ONE)
+			if (port == I2C_NUM_1) {
+				conf.sda_io_num = CONFIG_I2C_MANAGER_1_SDA;
+				conf.scl_io_num = CONFIG_I2C_MANAGER_1_SCL;
+				conf.sda_pullup_en = I2C_MANAGER_1_PULLUPS ? GPIO_PULLUP_ENABLE : GPIO_PULLUP_DISABLE;
+				conf.scl_pullup_en = conf.sda_pullup_en;
+				conf.master.clk_speed = CONFIG_I2C_MANAGER_1_FREQ_HZ;
+			}
+		#endif
+
+		conf.mode = I2C_MODE_MASTER;
+
+		ret = i2c_param_config(port, &conf);
+		ret |= i2c_driver_install(port, conf.mode, 0, 0, 0);
+
+		if (ret != ESP_OK) {
+			ESP_LOGE(TAG, "Failed to initialise I2C port %d.", (int)port);
+			ESP_LOGW(TAG, "If it was already open, we'll use it with whatever settings were used "
+			              "to open it. See I2C Manager README for details.");
+		} else {
+			ESP_LOGI(TAG, "Initialised port %d (SDA: %d, SCL: %d, speed: %d Hz.)",
+					 port, conf.sda_io_num, conf.scl_io_num, conf.master.clk_speed);
+		}
+
+	}
+
+    return ret;
+}
+
+esp_err_t I2C_FN(_read)(i2c_port_t port, uint16_t addr, uint32_t reg, uint8_t *buffer, uint16_t size) {
+
+	I2C_PORT_CHECK(port, ESP_FAIL);
+
+    esp_err_t result;
+
+    // May seem weird, but init starts with a check if it's needed, no need for that check twice.
+	I2C_FN(_init)(port);
+
+   	ESP_LOGV(TAG, "Reading port %d, addr 0x%03x, reg 0x%04x", port, addr, reg);
+
+	TickType_t timeout = 0;
+	#if defined (I2C_ZERO)
+		if (port == I2C_NUM_0) {
+			timeout = (CONFIG_I2C_MANAGER_0_TIMEOUT) / portTICK_RATE_MS;
```

Comment:
Could we use macros defined on lines 69 and 70?

Reply:
Yes, sorry, had missed this one...